docs/proposal/20220929-ipam-support.md

```diff
@@ -0,0 +1,423 @@
+  future work.
+
+## Proposal
+
+Following the contract specified by the IPAM integration proposal we plan to:
+
+- Add the addressesFromPools field to the spec.network.devices section of the
+  VSphereMachine and VSphereMachineTemplate. The addressesFromPools field will allow a
+  user to specify that the network device IPAddress should be claimed from an
+  IPPool.
+- If a VSphereMachine has been specified to claim an IP Address from an IPPool,
+  then CAPV will create an IPAddressClaim, which the IPAM provider will use to
+  allocate an IP Address. This IPAddress can then be used by the
+  VSphereMachine.
+
+### User Stories
+
+#### Story 1
+
+As a user, I want to be able to create a VSphereMachine where the IP Address
+for the machine comes from an IPAM Provider instead of DHCP.
+
+#### Story 2
+
+As a user I want to be able to allocate either IPv4 or IPv6 addresses for my
+VSphereMachine.
+
+#### Story 3
+
+As a user I want to be able to release my IP Address allocation when the
+VSphereMachine is deleted for future VSphereMachines to use.
+
+### Implementation Details/Notes/Constraints
+
+#### API Changes
+
+Adding `addressesFromPools`(type []corev1.TypedLocalObjectReference) to the VSphereMachineTemplate
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: example
+  namespace: vsphere-site1
+spec:
+  template:
+    spec:
+      cloneMode: FullClone
+      numCPUs: 8
+      memoryMiB: 8192
+      diskGiB: 45
+      network:
+        devices:
+        - dhcp4: false
+          dhcp6: false
+          networkName: VM Network
+          addressesFromPools:
+            # reference to the pool
+            - apiGroup: ipam.cluster.x-k8s.io
+              kind: InClusterIPPool
+              name: testpool4
+            - apiGroup: ipam.cluster.x-k8s.io
+              kind: InClusterIPPool
+              name: testpool6
+```
+
+Add `FromPools` to the NetworkDeviceSpec:
+
+```go
+// NetworkDeviceSpec defines the network configuration for a virtual machine's
+// network device.
+type NetworkDeviceSpec struct {
+ // FromPools is a list of references to the pools from which IP addresses should be claimed.
+ FromPools []corev1.TypedLocalObjectReference `json:"addressesFromPools,omitempty"`
+}
+```
+
+#### VSphereVM Controller Changes
+
+1. If the FromPools field is set on a device then this controller will create a
+   new IPAddressClaim for each pool configured on each device. The
+   IPAddressClaim will be named according to the device index and the pool
+   index in the VSphereVM spec.
+   The controller will not continue the VM creation process until all
+   IPAddressClaims have been fulfilled.
+
+    ```yaml
+    apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+    kind: IPAddressClaim
+    metadata:
+      name: "example-claim-0-0"
+      namespace: "vsphere-site1"
+      finalizers: ["ipam.cluster.x-k8s.io/ip-claim-protection"]

```

Comment:
What was the consideration around setting a finalizer? From what I have seen in the CAP* controllers, the controllers that reconcile the objects are the one setting and removing the finalizers. And I was under the assumption, that the IPAM provider implementation would have controllers that watch and add the finalizers.

Reply:
The IPAM provider will also set a finalized to free up the IP after the claim is removed.

The finalizer set by the infra provider is to prevent deletion of the claim while the IP is still in use by the Machine.

Reply:
We've updated the 'Owner Reference and Finalizer' section. Let us know if there's still more to explain.